### PR TITLE
feat: storing thumbnails on disk

### LINKED ISSUE #200: feat: storing thumbnails on disk
I don't know if this is completely outside the scope of the project, but I did a quick and dirty hack that caches thumbnails on disk similarly to how nsxiv does it. I ran some benchmarks and it saves 10 to 20 ms per image. The gallery still takes a very long time to load but the difference is there. If this is something that might get merged, I'd be happy to make a PR.

### PR DESCRIPTION
**Pull request by Rentib**
Closes #200

### CODE REVIEW

File: src/image.c [hunk 1]
```diff
@@ -65,6 +66,10 @@ void image_thumbnail(struct image* image, size_t size, bool fill,
     size_t thumb_height = scale * full->height;
     ssize_t offset_x, offset_y;
     enum pixmap_scale scaler;
+    static char* cache_dir = NULL;
+    char thumb_path[4096] = { 0 }; /* PATH_MAX */
```
Comment (artemsen) [inline src/image.c:70]: `PATH_MAX` is defined in `limits.h`, we can use it instead of magic numbers.

File: src/image.c [hunk 2]
```diff
@@ -82,12 +87,53 @@ void image_thumbnail(struct image* image, size_t size, bool fill,
         offset_y = 0;
     }
 
+    // TODO: probably should be moved outside of this function
```
Comment (artemsen) [inline src/image.c:90]: I agree =)

File: src/image.c [hunk 3]
```diff
@@ -82,12 +87,53 @@ void image_thumbnail(struct image* image, size_t size, bool fill,
...
+    if (!cache_dir) {
+        cache_dir = getenv("XDG_CACHE_HOME");
+        if (cache_dir) {
+            cache_dir = strcat(cache_dir, "/swayimg");
+        } else {
+            cache_dir = getenv("HOME");
+            if (!cache_dir) {
+                goto thumb_create;
+            }
+            cache_dir = strcat(cache_dir, "/.swayimg");
```
Comment (artemsen) [inline src/image.c:100]: `HOME/.cache/swayimg`, but it would be better to have an option in the config file so the user could set it themselves.

File: src/image.c [hunk 4]
```diff
@@ -82,12 +87,53 @@ void image_thumbnail(struct image* image, size_t size, bool fill,
...
+        }
+    }
+    sprintf(thumb_path, "%s%s", cache_dir, image->source);
+
+    if (!pixmap_load(&thumb, thumb_path)) {
+        goto thumb_create;
+    }
+
+    if (!(thumb.width == thumb_width && thumb.height == thumb_height)) {
```
Comment (artemsen) [inline src/image.c:109]: What if the file has been changed but the image size remains the same?
In this case, we need to recreate the thumbnail.

File: src/image.c [hunk 5]
```diff
@@ -82,12 +87,53 @@ void image_thumbnail(struct image* image, size_t size, bool fill,
...
     // create thumbnail
     if (!pixmap_create(&thumb, thumb_width, thumb_height)) {
         return;
     }
     pixmap_scale(scaler, full, &thumb, offset_x, offset_y, scale, image->alpha);
 
+    // save thumbnail to disk
+
+    // TODO: probably should be moved outside of this function
+    // TODO: do not use system() here
```
Comment (artemsen) [inline src/image.c:124]: Definitely.

File: src/pixmap.c [hunk 6]
```diff
@@ -300,6 +301,70 @@ void pixmap_free(struct pixmap* pm)
...
+bool pixmap_save(struct pixmap* pm, const char* path)
+{
+    FILE* fp;
+    unsigned i;
+
+    if (!(fp = fopen(path, "wb"))) {
+        return false;
+    }
+
+    // TODO: add alpha channel
```
Comment (artemsen) [inline src/pixmap.c:313]: We have 14 supported image formats. It would be better to use one of them instead of reinventing own format.

**Review by artemsen — CHANGES_REQUESTED** [review]
I think this feature should be optional. We need an option in the config file that will allow saving thumbnails to disk. Writing anything to persistent storage by default is not a good idea.

File: src/image.c (hunk 3 above)
Comment (Rentib) [reply src/image.c]: Adding this to config would be better for another PR i think, but I can try adding it here too...

File: src/image.c (hunk 4 above)
Comment (Rentib) [reply src/image.c]: I have no idea how to check this to be honest. Some libnotify thing?

File: src/pixmap.c (hunk 6 above)
Comment (Rentib) [reply src/pixmap.c:313]: I don't see any function for writing an image to disk in the codebase. Did I miss something?

File: src/image.c (hunk 3 above)
Comment (artemsen) [reply src/image.c]: Yes, this can be taken out into a separate request.
But in any case, this PR should contain the configurable option for enabling/disabling the persistent cache.

File: src/image.c (hunk 4 above)
Comment (artemsen) [reply src/image.c]: I would compare mtime of the files.

File: src/pixmap.c (hunk 6 above)
Comment (artemsen) [reply src/pixmap.c:313]: Current code doesn't contain code for encoders. But these APIs are available to `swayimg`.

File: src/image.c [hunk 7]
```diff
@@ -4,9 +4,13 @@
 
 #include "image.h"
 
+#include <errno.h>
+#include <linux/limits.h>
```
Comment (artemsen) [inline src/image.c:8]: This breaks FreeBSD build.

File: src/image.c [hunk 8]
```diff
@@ -51,6 +55,63 @@ void image_rotate(struct image* ctx, size_t angle)
...
+    return r >= 0 && r < PATH_MAX;
+}
+
+// TODO: where should this function be?
+/**
+ * Makes directories like `mkdir -p`.
+ * @param path absolute path to the directory to be created
+ * @return true if successful
+ */
+static bool make_directories(char* path)
```
Comment (artemsen) [inline src/image.c:91]: `make_directory` in image entity. Something went wrong in program architecture =)

File: src/pixmap.c [hunk 9]
```diff
@@ -300,6 +301,70 @@ void pixmap_free(struct pixmap* pm)
     free(pm->data);
 }
 
+bool pixmap_save(struct pixmap* pm, const char* path)
+{
+    FILE* fp;
+    unsigned i;
```
Comment (artemsen) [inline src/pixmap.c:307]: Please, use fixed-size types.

File: src/image.c [hunk 10]
```diff
@@ -51,6 +55,63 @@ void image_rotate(struct image* ctx, size_t angle)
...
+{
+    char* slash;
+
+    if (!*path) {
+        return true;
```
Comment (artemsen) [inline src/image.c:96]: Why this function may be called for empty directory?

File: src/image.c [hunk 11]
```diff
@@ -51,6 +55,63 @@ void image_rotate(struct image* ctx, size_t angle)
...
+
+    slash = strchr(slash + 1, '/');
+    while (slash) {
+        *slash = '\0';
+        if (mkdir(path, 0755) && errno != EEXIST) {
+            // TODO: do we want to print error message?
+            return false;
+        }
+        *slash = '/';
+        slash = strchr(slash + 1, '/');
```
Comment (artemsen) [inline src/image.c:109]: Copy/paste from line 101.

File: src/image.c (hunk 4 above)
Comment (Rentib) [reply src/image.c]: Then mtime would have to be stored somehow in the thumbnail or another file with mtimes of cached images.
Is there some better method for this?

File: src/image.c [hunk 12]
```diff
@@ -51,6 +55,63 @@ void image_rotate(struct image* ctx, size_t angle)
...
+
+    if (!*path) {
+        return true;
+    }
+
+    slash = path;
+
+    slash = strchr(slash + 1, '/');
+    while (slash) {
+        *slash = '\0';
```
Comment (artemsen) [inline src/image.c:103]: It's a bad practice. You modify the input buffer, that should be const.

File: src/image.c [hunk 13]
```diff
@@ -82,12 +144,25 @@ void image_thumbnail(struct image* image, size_t size, bool fill,
         offset_y = 0;
     }
 
+    // get thumbnail from cache
+    if (get_image_thumb_path(image, thumb_path) &&
```
Comment (artemsen) [inline src/image.c:148]: `image_thumbnail` is intended to create a thumbnail, not load.
Any function should do one thing =)

File: src/image.c [hunk 14]
```diff
@@ -51,6 +55,63 @@ void image_rotate(struct image* ctx, size_t angle)
...
+            cache_dir = strcat(cache_dir, "/swayimg");
+        } else {
+            cache_dir = getenv("HOME");
+            if (!cache_dir) {
+                return false;
+            }
+            cache_dir = strcat(cache_dir, "/.swayimg");
+        }
+    }
+    r = snprintf(path, PATH_MAX, "%s%s", cache_dir, image->source);
```
Comment (artemsen) [inline src/image.c:81]: `snprintf(path, sizeof(path), ...`

File: src/image.c [hunk 15]
```diff
@@ -51,6 +55,63 @@ void image_rotate(struct image* ctx, size_t angle)
...
+ * @return true if successful
+ */
+static bool get_image_thumb_path(struct image* image, char* path)
+{
+    static char* cache_dir = NULL;
+    int r;
+    if (!cache_dir) {
+        cache_dir = getenv("XDG_CACHE_HOME");
+        if (cache_dir) {
+            cache_dir = strcat(cache_dir, "/swayimg");
```
Comment (artemsen) [inline src/image.c:72]: Wow =)

File: src/pixmap.c [hunk 16]
```diff
@@ -300,6 +301,70 @@ void pixmap_free(struct pixmap* pm)
...
+        return false;
+    }
+
+    char header[3];
+    if (fscanf(fp, "%2s\n%zu %zu\n255\n", header, &pm->width, &pm->height) !=
+        3) {
+        goto fail;
+    }
+
+    if (strcmp(header, "P6")) {
```
Comment (artemsen) [inline src/pixmap.c:341]: Just curious, what does "P6" mean?

File: src/pixmap.c [hunk 17]
```diff
@@ -300,6 +301,70 @@ void pixmap_free(struct pixmap* pm)
...
+    unsigned i;
+
+    if (!(fp = fopen(path, "wb"))) {
+        return false;
+    }
+
+    // TODO: add alpha channel
+    fprintf(fp, "P6\n%zu %zu\n255\n", pm->width, pm->height);
+    for (i = 0; i < pm->width * pm->height; ++i) {
+        uint8_t color[4] = { (((pm->data[i] >> (8 * 2)) & 0xff)),
```
Comment (artemsen) [inline src/pixmap.c:316]: 4-items array, but filled only 3.

**Review by artemsen — CHANGES_REQUESTED** [review]

File: src/image.c (hunk 4 above)
Comment (artemsen) [reply src/image.c]: Why not use the file system for this? That's what it's designed for =)

Comment (Rentib) [reply src/image.c]: Ah, I get it. Compare the mtime of thumbnail and image and if image's mtime is greater than remake the thumbnail, right?

File: src/image.c (hunk 11 above)
Comment (Rentib) [reply src/image.c]: I don't really see anything wrong with it, but I changed it.

File: src/image.c (hunk 8 above)
Comment (Rentib) [reply src/image.c:91]: I don't know where to put it...

File: src/image.c (hunk 13 above)
Comment (Rentib) [reply src/image.c:154]: Should I change gallery.c: add_thumbnail() to first try to get thumbnail from cache and if it is unavailable, then run image_thumbnail() and store it on disk?

File: src/image.c (hunk 4 above)
Comment (Rentib) [reply src/image.c]: It must be ctime

File: src/image.c (hunk 8 above)
Comment (artemsen) [reply src/image.c:101]: I would create `thumbnail.c` and move all these changes there (include pixman).

File: src/pixmap.c (hunk 16 above)
Comment (artemsen) [reply src/pixmap.c:341]: No, [it's not](https://github.com/artemsen/swayimg/blob/master/src/formats/bmp.c).
But you can add real BMP encoder for that =).

File: src/image.c (hunk 4 above)
Comment (artemsen) [reply src/image.c]: One more thing: if antialising mode or fill mode were changed, then we have to recreate thumbnails too.

Comment (Rentib) [reply src/image.c]: This makes sense, do you have any idea on how to check this?
I could maybe look at config's ctime and recreate each time it is changed, but I don't really like this idea.

Comment (artemsen) [reply src/image.c]: Store this data within the thumbnail? In meta/exif/...

File: src/pixmap.c (hunk 16 above)
Comment (Rentib) [reply src/pixmap.c]: Ah sorry, my mistake. It's PPM format. Probably the simplest thing out there.

File: src/image.c (hunk 4 above)
Comment (Rentib) [reply src/image.c]: done

File: src/image.c (hunk 15 above)
Comment (Rentib) [reply src/image.c]: Would it be fine to put `export_path()` in config.h?
Then I could use it here

File: src/pixmap.c (hunk 6 above)
Comment (artemsen) [reply src/pixmap.c]: I think it would be better to add `encode_pnm` to [PNM(PPM)](https://github.com/artemsen/swayimg/blob/master/src/formats/pnm.c) decoder.

**Comment by artemsen** [comment]
Please, rebase to latest master. This will fix Ubuntu CI.

File: src/image.c (hunk 15 above)
Comment (Rentib) [reply src/image.c]: Sorry, I meant `expand_path()`.

Comment (artemsen) [reply src/image.c]: This function is specific to thumbnails, so leave it in `thumbnail.c`.

Comment (Rentib) [reply src/image.c]: Let me rephrase the question:
Would it be alright to add `expand_path()` to config.h header file?
It is currently implemented as a static function in config.c:203

Comment (artemsen) [reply src/image.c]: Ah, I get it =)
Yes, it will be ok.

File: src/thumbnail.c [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    if (!(fp = fopen(path, "rb"))) {
+        goto fail;
+    }
+
+    char header[3];
+    if (fscanf(fp, "%2s\n%zu %zu\n255\n", header, &thumb->width,
+               &thumb->height) != 3) {
+        goto fail;
```
Comment (artemsen) [inline src/thumbnail.c:179]: You love `goto` so much =)
Maybe there is another way to handle errors?

**Review by artemsen — CHANGES_REQUESTED** [review]

File: src/thumbnail.c [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+#elif defined(__OpenBSD__) || defined(__FreeBSD__)
+#define PATH_MAX 4096
+#endif
+
+/**
+ * Makes directories like `mkdir -p`.
+ * @param path absolute path to the directory to be created
+ * @return true if successful
+ */
+static bool make_directories(char* path)
```
[the remark this answers is not in the document]
Comment (Rentib) [reply src/thumbnail.c:26]: Then I would need to strdup it

File: src/thumbnail.c [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+{
+    char* slash;
+
+    if (!path || !*path) {
```
[the remark this answers is not in the document]
Comment (Rentib) [reply src/thumbnail.c:30]: Yes, but isn't it better to be safe? I can remove the condition, but it doesn't really impact the performance.

File: src/thumbnail.c [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+        if (mkdir(path, 0755) && errno != EEXIST) {
+            return false;
+        }
+        *slash = '/';
+    }
+
+    return true;
+}
+
+static bool get_thumb_path(char** path, const char* source)
```
[the remark this answers is not in the document]
Comment (Rentib) [reply src/thumbnail.c:50]: Ok, this is a problem. I will fix it to not pass source, but the absolute file path.

File: src/thumbnail.h [hunk 22]
```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: MIT
+// Image instance: pixel data and meta info.
+// Copyright (C) 2021 Artem Senichev <artemsen@gmail.com>
```
[the remark this answers is not in the document]
Comment (Rentib) [reply src/thumbnail.h:3]: Function `thumbnail_create()` is your work (I copy/pasted). I can remove the copyright if you want though.

File: src/thumbnail.c (hunk 18 above)
Comment (Rentib) [reply src/thumbnail.c:176]: This is the same way of handling errors as in `formats/avif.c` and many other files in the project. I could maybe make a macro which cleans stuff up and then returns false, but I believe this is cleaner.

File: src/thumbnail.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (artemsen) [reply src/thumbnail.c:30]: It is not about performance, but about architecture. Why do we have cases when path is NULL or empty?

File: src/thumbnail.c [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+{
+    char* slash;
+
+    if (!path || !*path) {
+        return false;
+    }
+
+    slash = path;
+    while (true) {
+        slash = strchr(slash + 1, '/');
```
[the remark this answers is not in the document]
Comment (Rentib) [reply src/thumbnail.c:36]: I don't see an easy way to concatenate slices. I need to mkdir a prefix of `path` every time.

**Comment by artemsen** [comment]
@Rentib, I appreciate your efforts and thanks for the prototype!
